src/deploy/functions/release/planner.ts

```diff
@@ -0,0 +1,236 @@
+    return false;
+  }
+  if (have.platform !== "gcfv2") {
+    return false;
+  }
+  if (!backend.isEventTriggered(want)) {
+    return false;
+  }
+  if (!backend.isEventTriggered(have)) {
+    return false;
+  }
+  if (want.eventTrigger.eventType != gcfv2.PUBSUB_PUBLISH_EVENT) {
+    return false;
+  }
+  if (have.eventTrigger.eventType !== gcfv2.PUBSUB_PUBLISH_EVENT) {
+    return false;
+  }
+  return have.eventTrigger.eventFilters["resource"] != want.eventTrigger.eventFilters["resource"];
+}
+
+/** Whether a user upgraded a scheduled function (which goes from Pub/Sub to HTTPS). */
+export function upgradedScheduleFromV1ToV2(
+  want: backend.Endpoint,
+  have: backend.Endpoint
+): boolean {
+  if (have.platform !== "gcfv1") {
+    return false;
+  }
+  if (want.platform !== "gcfv2") {
+    return false;
+  }
+  if (!backend.isScheduleTriggered(have)) {
+    return false;
+  }
+  // should not be possible
+  if (!backend.isScheduleTriggered(want)) {
+    return false;
+  }
+
+  return true;
+}
+
+/** Throws if there is an illegal update to a function. */
+export function checkForIllegalUpdate(want: backend.Endpoint, have: backend.Endpoint): void {
+  const triggerType = (e: backend.Endpoint): string => {
+    if (backend.isHttpsTriggered(e)) {
+      return "an HTTPS";
+    } else if (backend.isEventTriggered(e)) {
+      return "a background triggered";
+    } else if (backend.isScheduleTriggered(e)) {
+      return "a scheduled";
+    }
+    // Unfortunately TypeScript isn't like Scala and I can't prove to it
+    // that all cases have been handled
+    throw Error("Functions release planner is not able to handle an unknown trigger type");
+  };
+  const wantType = triggerType(want);
+  const haveType = triggerType(have);
+  if (wantType != haveType) {
+    throw new FirebaseError(
+      `[${getFunctionLabel(
+        want
+      )}] Changing from ${haveType} function to ${haveType} function is not allowed. Please delete your function and create a new one instead.`
+    );
+  }
+  if (want.platform == "gcfv1" && have.platform == "gcfv2") {
+    throw new FirebaseError(
+      `[${getFunctionLabel(want)}] Functions cannot be downgraded from GCFv2 to GCFv1`
+    );
+  }
+
+  // We need to call from module exports so tests can stub this behavior, but that
+  // breaks the type system.
+  // eslint-disable-next-line
+  exports.checkForV2Upgrade(want, have);
+}
+
+/**
+ * Throws an error when upgrading/downgrading GCF versions.
+ * This is a separate function that is designed to be stubbed in tests to allow
+ * upgrading to v2 in tests before production is ready
+ */
+export function checkForV2Upgrade(want: backend.Endpoint, have: backend.Endpoint): void {
+  if (want.platform == "gcfv2" && have.platform == "gcfv1") {
+    throw new FirebaseError(
+      `[${getFunctionLabel(
+        have
+      )}] Upgrading from GCFv1 to GCFv2 is not yet supported. Please delete your old function or wait for this feature to be ready.`
```

Comment:
I'm wondering if we should use `CF3v{1/2}` instead of `GCFv{1,2}`? Is `CF3v2 === GCFv2` always going to hold or could `CF3v2` target platform other than `GCFv2`?

Reply:
CF3v2 is an SDK, GCFv2 is a platform. It's possible we would have a different `platform` from the same SDK, just like we're going to support `platform: "gcfv1"` in CF3v1 (even if begrudgingly). So I think GCFv1/2 is the right label here.